### DISCUSSION TITLE
What's the recommended method of calling a label printer from a plugin?

### DISCUSSION THREAD

**Discussion by miggland** [inventree/InvenTree#4331] [Q&A]
I want to print a label from a plugin.

That is, my plugin should only call the 'print' function, and tell InvenTree to use a specific, installed label printer plugin to run the printing - basically, I want the same effect as if calling an URLlike this:
`/api/label/stock/1/print/?items[]=408&plugin=printer`
from within my plugin.

But how do I call this from within Python in a plugin?

**Comment by SchrodingersGat** [comment]
@miggland good question. Check out what happens in the backend when you call that API URL:

https://github.com/inventree/InvenTree/blob/f4e8c05165aeb9bcfb395830f66b353559683fac/InvenTree/label/api.py#L230

**Reply by miggland** [reply]
OK, I guess that would work. However, I was hoping there was something like the "print" function which I could use directly. In this case, I don't know what to send as the request though..

**Reply by SchrodingersGat** [reply]
@miggland it is all integrated pretty closely to the plugin architecture, but you could extract the bits you need to do it manually, if you need.

With regard to the "request" object - try passing `None` and see if you can make it work? It's been a while since I played around with this code, sorry :)

**Reply by miggland** [reply]
@SchrodingersGat I'm trying to do this now.

The problem is not the integration with the plugin architecture, but that creating labels seems to rely heavily on the request sent. A call to `request` is present in a lot of places, for generating context, filenames, etc. Even WeasyPrint relies on the request data..

```
si = stock.StockItem.objects.get(pk=kwargs['id'])
labelitem = StockItemLabel(pk=self.get_setting('STOCK_LABEL_ITEM'))
labelitem.object_to_print = si

offload_task(
                    plugin_label.print_label,
                    self.get_setting('PRINTER'),
                    pdf,
                    filename='label.pdf',
                    label_instance=labelitem,
                    user=None,
)
```

What I'm missing is the actual PDF data, `pdf`.

It should be accessible from `labelitem.render(request).get_document().write_pdf()`

Putting None here gets me further if I change a few functions in label.models (see below), but then once WeasyPrint is called, there are still requests which need to be satisfied..

Getting all these items fixed is maybe possible, but if I could just call the correct URL I wouldn't have to care about all these details 


Error:
File "/home/inventree/data/env/lib/python3.9/site-packages/django_q/cluster.py", line 432, in worker
```
inventree-dev-worker    |     res = f(*task["args"], **task["kwargs"])
inventree-dev-worker    |   File "/home/inventree/InvenTree/plugin/base/event/events.py", line 94, in process_event
inventree-dev-worker    |     plugin.process_event(event, *args, **kwargs)
inventree-dev-worker    |   File "/home/inventree/InvenTree/plugins/autoprintlabel/AutoPrintLabelPlugin.py", line 215, in process_event
inventree-dev-worker    |     labelitem.render(None).get_document().write_pdf(),
inventree-dev-worker    |   File "/home/inventree/data/env/lib/python3.9/site-packages/django_weasyprint/views.py", line 71, in get_document
inventree-dev-worker    |     base_url = self.get_base_url()
inventree-dev-worker    |   File "/home/inventree/data/env/lib/python3.9/site-packages/django_weasyprint/views.py", line 34, in get_base_url
inventree-dev-worker    |     self._request.build_absolute_uri('/')
inventree-dev-worker    | AttributeError: 'NoneType' object has no attribute 'build_absolute_uri'
```



Changes, for testing purposes:
```
--- a/InvenTree/label/models.py
+++ b/InvenTree/label/models.py
@@ -185,7 +185,7 @@ class LabelTemplate(models.Model):
         context['date'] = datetime.datetime.now().date()
         context['datetime'] = datetime.datetime.now()
         context['request'] = request
-        context['user'] = request.user
+        context['user'] = request.user if request else None
         context['width'] = self.width
         context['height'] = self.height
 
@@ -206,9 +206,9 @@ class LabelTemplate(models.Model):
         wp = WeasyprintLabelMixin(
             request,
             self.template_name,
-            base_url=request.build_absolute_uri("/"),
+            base_url='', #request.build_absolute_uri("/"),
             presentational_hints=True,
-            filename=self.generate_filename(request),
+            filename='label.pdf', #self.generate_filename(request),
             **kwargs
         )
 
@@ -252,7 +252,7 @@ class StockItemLabel(LabelTemplate):
             'barcode_data': stock_item.barcode_data,
             'barcode_hash': stock_item.barcode_hash,
             'qr_data': stock_item.format_barcode(brief=True),
-            'qr_url': request.build_absolute_uri(stock_item.get_absolute_url()),
+            'qr_url': request.build_absolute_uri(stock_item.get_absolute_url()) if request else stock_item.get_absolute_url(),
             'tests': stock_item.testResultMap(),
             'parameters': stock_item.part.parameters_map(),
```

**Reply by matmair** [reply]
@miggland getting a request object is not that difficult; my preffered way is using a [factory](https://docs.djangoproject.com/en/4.1/topics/testing/advanced/#django.test.RequestFactory) but you could also use the test client or implement a first party call structure that does not need a request. We need the request mostly to generate a context (what object is requested and by which user on which domain) - we could serialize the values eraly on and then use that entrypoint for non-request calls of the trace.

**Reply by miggland** [reply]
That's the solution I was looking for!

My (still clunky, but at least working solution now):
```
final_url = f"{reverse('api-stockitem-label-print', kwargs={'pk': self.get_setting('STOCK_LABEL_ITEM')})}/?items[]={kwargs['id']}&plugin={self.get_setting('PRINTER')}"

# Use RequestFactory to create a request object which can be sent to the API view
factory = RequestFactory()
pluginrequest = factory.get(final_url)
# Fake the admin user here for now - there may be a better solution
pluginrequest.user = User.objects.get(pk=1)

response = StockItemLabelPrint.as_view()(pluginrequest, pk=self.get_setting('STOCK_LABEL_ITEM'))
```

Thanks for the help!
